### ISSUE TITLE
Memory not aligned when using GPU on Intel Arc (Intel Ultra 185H)

### ISSUE THREAD

**Issue by danielgalek** [intel/intel-extension-for-tensorflow#78]
My machine is recognizing the XPU, is able to run the samples but keeps crashing with the error:

2024-09-07 23:53:11.032822: I tensorflow/core/common_runtime/next_pluggable_device/next_pluggable_device_factory.cc:118] Created 1 TensorFlow NextPluggableDevices. Physical device type: XPU
2024-09-07 23:53:41.345227: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:117] Plugin optimizer for device_type XPU is enabled.
2024-09-07 23:53:59.266808: F tensorflow/core/framework/tensor.cc:847] Check failed: IsAligned() ptr = 0x7e3664005e51
Aborted (core dumped)

The code works fine when running on intel-extension-for-tensorflow on CPU, it works fine without the extension and it works fine with NVIDIA  T4 tensor GPU.

My setup:  
Intel® Core™ Ultra processors (Series 1) 185H
Integrated Intel Arc Graphics 2.35 GHz, 8 Xe-Cores 
RAM: 64GB
I tried WSL2 (ubuntu 22.04) on Win 11 Pro and directly on ubuntu 22.04 directly (dualboot), always with the same result)

I tried many things like ITEX_LIMIT_MEMORY_SIZE_IN_MB, TF_ENABLE_ONEDNN_OPTS=0, set_memory_growth, stable and nightly builds of inte-extension-for-tensorflow etc. always with the same error, Once in a while this one also appears before the IsAligned() crash: 2024-09-07 23:47:26.258971: W external/local_tsl/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 491765760 exceeds 10% of free system memory.

env_check returns this:

Check Environment for Intel(R) Extension for TensorFlow*...

Check Python
	 Python 3.10.12 is Supported.
Check Python Passed

Check OS
	OS ubuntu:22.04 is Supported
Check OS Passed

Check Tensorflow
	Tensorflow 2.15.1 is installed.
Check Tensorflow Passed

Check Intel GPU Driver
Package: intel-level-zero-gpu
Status: install ok installed
Priority: optional
Section: libs
Installed-Size: 28239
Maintainer: Intel Graphics Team <linux-graphics@intel.com>
Architecture: amd64
Source: intel-compute-runtime
Version: 1.3.27642.52-803 22.04
Depends: libc6 (>= 2.34), libgcc-s1 (>= 3.4), libigdgmm12 (>= 22.3.15), libstdc++6 (>= 12), libigc1 (>= 1.0.12812), libigdfcl1 (>= 1.0.12812), libnl-3-200, libnl-route-3-200
Description: Intel(R) Graphics Compute Runtime for oneAPI Level Zero.
 Level Zero is the primary low-level interface for language and runtime
 libraries. Level Zero offers fine-grain control over accelerators
 capabilities, delivering a simplified and low-latency interface to
 hardware, and efficiently exposing hardware capabilities to applications.
Homepage: https://github.com/oneapi-src/level-zero
Original-Maintainer: Debian OpenCL Maintainers <pkg-opencl-devel@lists.alioth.debian.org>
Package: intel-opencl-icd
Status: install ok installed
Priority: optional
Section: libs
Installed-Size: 23865
Maintainer: Intel Graphics Team <linux-graphics@intel.com>
Architecture: amd64
Source: intel-compute-runtime
Version: 23.43.27642.52-803~22.04
Replaces: intel-opencl
Provides: opencl-icd
Depends: libc6 (>= 2.34), libgcc-s1 (>= 3.4), libigdgmm12 (>= 22.3.15), libstdc++6 (>= 12), ocl-icd-libopencl1, libigc1 (>= 1.0.12812), libigdfcl1 (>= 1.0.12812)
Recommends: intel-igc-cm (>= 1.0.100)
Breaks: intel-opencl
Conffiles:
 /etc/OpenCL/vendors/intel.icd d0a34d0b4f75385c56ee357bb1b8e2d0
Description: Intel graphics compute runtime for OpenCL
 The Intel(R) Graphics Compute Runtime for OpenCL(TM) is a open source
 project to converge Intel's development efforts on OpenCL(TM) compute
 stacks supporting the GEN graphics hardware architecture.
 .
 Supported platforms:
 - Intel Core Processors with Gen8 GPU (Broadwell) - OpenCL 2.1
 - Intel Core Processors with Gen9 GPU (Skylake, Kaby Lake, Coffee Lake) - OpenCL 2.1
 - Intel Atom Processors with Gen9 GPU (Apollo Lake, Gemini Lake) - OpenCL 1.2
 - Intel Core Processors with Gen11 GPU (Ice Lake) - OpenCL 2.1
 - Intel Core Processors with Gen12 graphics devices (formerly Tiger Lake) - OpenCL 2.1
Homepage: https://github.com/intel/compute-runtime
Original-Maintainer: Debian OpenCL Maintainers <pkg-opencl-devel@lists.alioth.debian.org>
Package: level-zero
Status: install ok installed
Priority: optional
Section: libs
Installed-Size: 1049
Maintainer: Intel Graphics Team <linux-graphics@intel.com>
Architecture: amd64
Source: level-zero-loader
Version: 1.14.0-744 22.04
Depends: libc6 (>= 2.34), libgcc-s1 (>= 3.3.1), libstdc++6 (>= 11)
Description: Intel(R) Graphics Compute Runtime for oneAPI Level Zero.
 Level Zero is the primary low-level interface for language and runtime
 libraries. Level Zero offers fine-grain control over accelerators
 capabilities, delivering a simplified and low-latency interface to
 hardware, and efficiently exposing hardware capabilities to applications.
 .
 This package provides the loader for oneAPI Level Zero compute runtimes.
Homepage: https://github.com/oneapi-src/level-zero
Package: libigc1
Status: install ok installed
Priority: optional
Section: libs
Installed-Size: 86364
Maintainer: Intel Graphics Team <linux-graphics@intel.com>
Architecture: amd64
Source: intel-graphics-compiler
Version: 1.0.15468.29-803 22.04
Depends: libc6 (>= 2.34), libgcc-s1 (>= 3.4), libstdc++6 (>= 12), zlib1g (>= 1:1.2.2)
Description: Intel graphics compiler for OpenCL -- core libs
 The Intel(R) Graphics Compiler for OpenCL(TM) is an llvm based compiler
 for OpenCL(TM) targeting Intel Gen graphics hardware architecture.
 .
 This package includes the core libraries.
Homepage: https://github.com/intel/intel-graphics-compiler
Original-Maintainer: Debian OpenCL team <pkg-opencl-devel@lists.alioth.debian.org>
Package: libigdfcl1
Status: install ok installed
Priority: optional
Section: libs
Installed-Size: 116046
Maintainer: Intel Graphics Team <linux-graphics@intel.com>
Architecture: amd64
Source: intel-graphics-compiler
Version: 1.0.15468.29-803 22.04
Depends: libc6 (>= 2.34), libgcc-s1 (>= 3.4), libstdc++6 (>= 11), zlib1g (>= 1:1.2.0), libz3-4 (>= 4.7.1)
Description: Intel graphics compiler for OpenCL -- OpenCL library
 The Intel(R) Graphics Compiler for OpenCL(TM) is an llvm based compiler
 for OpenCL(TM) targeting Intel Gen graphics hardware architecture.
 .
 This package includes the library for OpenCL.
Homepage: https://github.com/intel/intel-graphics-compiler
Original-Maintainer: Debian OpenCL team <pkg-opencl-devel@lists.alioth.debian.org>
Package: libigdgmm12
Status: install ok installed
Priority: optional
Section: libs
Installed-Size: 648
Maintainer: Intel Graphics Team <linux-graphics@intel.com>
Architecture: amd64
Multi-Arch: same
Source: intel-gmmlib
Version: 22.3.15-803 22.04
Replaces: libigdgmm11
Depends: libc6 (>= 2.34), libgcc-s1 (>= 3.3.1), libstdc++6 (>= 4.1.1)
Description: Intel Graphics Memory Management Library -- shared library
 The Intel Graphics Memory Management Library provides device specific
 and buffer management for the Intel Graphics Compute Runtime for
 OpenCL and the Intel Media Driver for VAAPI.
 .
 This library is only useful for Broadwell and newer CPUs.
 .
 This package includes the shared library.
Homepage: https://github.com/intel/gmmlib
Original-Maintainer: Debian Multimedia Maintainers <debian-multimedia@lists.debian.org>
Check Intel GPU Driver Passsed

Check OneAPI
	Can't find dpcpp
 Check OneAPI Failed
 
 However, i do have OneApi installed, I tried 2024.1 and 2024.2. Same result.

**Comment by wangkl2** [comment]
@danielgalek Thanks for reporting this issue.
First, as in [install_for_xpu](https://github.com/intel/intel-extension-for-tensorflow/blob/main/docs/install/install_for_xpu.md#hardware-requirements), ITEX XPU officially supports Intel Data Center GPUs including Max and Flex series, but has `experimental support` for the `Client Arc-A series`, whatever for the `dGPU` or the `iGPU (such as Arc iGPU in Core Ultra)`.

For the issue you met, from the output log of the env_check tool, there is no output revealing whether XPU is detected. `Can't find dpcpp` might be caused by you've installed the `runtime` version of dpcpp/mkl instead of the full compiler/mkl. To make it work, please try install the full oneAPI as in [optional-install-full-intel-oneapi](https://github.com/intel/intel-extension-for-tensorflow/blob/main/docs/install/experimental/install_for_arc_gpu.md#optional-install-full-intel-oneapi). 

And please also share the output of `python -c "import intel_extension_for_tensorflow as itex; print(itex.__version__)"` for sanity check whether the XPU device is detected by ITEX.

**Comment by danielgalek** [comment]
Thank you @wangkl2 for your fast reply. I am aware that the support is experimental but I thought maybe by submitting issues might help.

I had installed the full oneAPI, first 2024.2, then while reading through some threads back to 2024.1 and now reinstalled 2024.2. Here is a full screenshot

![Screenshot from 2024-09-09 14-29-58](https://github.com/user-attachments/assets/fc11b362-8cbe-4aac-a198-f9eaafad73de)

Or is there any other way to check for the compiler/mkl version? 

I assume that the XPU is detected:

2024-09-07 23:53:11.032822: I tensorflow/core/common_runtime/next_pluggable_device/next_pluggable_device_factory.cc:118] **Created 1 TensorFlow NextPluggableDevices. Physical device type: XPU**
2024-09-07 23:53:41.345227: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:117] **Plugin optimizer for device_type XPU is enabled.**

here is the output from python -c "import intel_extension_for_tensorflow as itex; print(itex.__version__)"

2024-09-09 14:26:28.612571: I tensorflow/core/util/port.cc:113] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-09-09 14:26:28.613848: I external/local_tsl/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
2024-09-09 14:26:28.633540: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-09-09 14:26:28.633569: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-09-09 14:26:28.634161: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-09-09 14:26:28.637568: I external/local_tsl/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
2024-09-09 14:26:28.637783: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-09-09 14:26:29.001118: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
2024-09-09 14:26:29.525783: W external/local_tsl/tsl/lib/monitoring/collection_registry.cc:81] Trying to register 2 metrics with the same name: /tensorflow/core/bfc_allocator_delay. The old value will be erased in order to register a new one. Please check if you link the metric more than once, or if the name is already used by other metrics.
2024-09-09 14:26:29.525885: W external/local_tsl/tsl/lib/monitoring/collection_registry.cc:81] Trying to register 2 metrics with the same name: /xla/service/gpu/compiled_programs_count. The old value will be erased in order to register a new one. Please check if you link the metric more than once, or if the name is already used by other metrics.
2024-09-09 14:26:29.526467: W external/local_tsl/tsl/lib/monitoring/collection_registry.cc:81] Trying to register 2 metrics with the same name: /jax/pjrt/pjrt_executable_executions. The old value will be erased in order to register a new one. Please check if you link the metric more than once, or if the name is already used by other metrics.
2024-09-09 14:26:29.526477: W external/local_tsl/tsl/lib/monitoring/collection_registry.cc:81] Trying to register 2 metrics with the same name: /jax/pjrt/pjrt_executable_execution_time_usecs. The old value will be erased in order to register a new one. Please check if you link the metric more than once, or if the name is already used by other metrics.
2024-09-09 14:26:29.709892: I itex/core/wrapper/itex_gpu_wrapper.cc:38] Intel Extension for Tensorflow* GPU backend is loaded.
2024-09-09 14:26:29.710157: I external/local_xla/xla/pjrt/pjrt_api.cc:67] PJRT_Api is set for device type xpu
2024-09-09 14:26:29.710165: I external/local_xla/xla/pjrt/pjrt_api.cc:72] PJRT plugin for XPU has PJRT API version 0.33. The framework PJRT API version is 0.34.
2024-09-09 14:26:29.722651: I external/intel_xla/xla/stream_executor/sycl/sycl_gpu_runtime.cc:134] Selected platform: Intel(R) Level-Zero
2024-09-09 14:26:29.722852: I external/intel_xla/xla/stream_executor/sycl/sycl_gpu_runtime.cc:159] number of sub-devices is zero, expose root device.
2024-09-09 14:26:29.723897: I external/xla/xla/service/service.cc:168] XLA service 0x625b82e02210 initialized for platform SYCL (this does not guarantee that XLA will be used). Devices:
2024-09-09 14:26:29.723932: I external/xla/xla/service/service.cc:176]   StreamExecutor device (0): Intel(R) Graphics [0x7d55], <undefined>
2024-09-09 14:26:29.725062: I itex/core/devices/gpu/itex_gpu_runtime.cc:130] Selected platform: Intel(R) Level-Zero
2024-09-09 14:26:29.725242: I itex/core/devices/gpu/itex_gpu_runtime.cc:155] number of sub-devices is zero, expose root device.
2024-09-09 14:26:29.725457: I external/intel_xla/xla/pjrt/se_xpu_pjrt_client.cc:97] Using BFC allocator.
2024-09-09 14:26:29.725468: I external/xla/xla/pjrt/gpu/gpu_helpers.cc:106] XLA backend allocating 56115744768 bytes on device 0 for BFCAllocator.
2024-09-09 14:26:29.726658: I external/local_xla/xla/pjrt/pjrt_c_api_client.cc:119] PjRtCApiClient created.

import tensorflow as tf
physical_devices = tf.config.list_physical_devices('XPU')
print(physical_devices)

returns

[PhysicalDevice(name='/physical_device:XPU:0', device_type='XPU')]

**Comment by wangkl2** [comment]
@danielgalek 
To check the installed compiler/mkl version, 
1) if you are installing them (runtime version) via apt as you did previously, you can use `dpkg -l |grep oneapi`;
2) if you are installing them (the complete version) via the oneapi online/offline installer, you can use `dpcpp -v` after `source /opt/intel/oneapi/setvars.sh`

The latest itex version `v2.15.0.1` works with oneapi base toolkit `2024.2`. 

Now from your output of `python -c "import intel_extension_for_tensorflow as itex; print(itex.version)"`, the Arc iGPU (with 0x7d55 as the device ID) can be detected via itex, the XPU backend can be successfully registered and loaded via the PJRT plugin API.
```
2024-09-09 14:26:29.709892: I itex/core/wrapper/itex_gpu_wrapper.cc:38] Intel Extension for Tensorflow* GPU backend is loaded.
2024-09-09 14:26:29.710157: I external/local_xla/xla/pjrt/pjrt_api.cc:67] PJRT_Api is set for device type xpu
2024-09-09 14:26:29.710165: I external/local_xla/xla/pjrt/pjrt_api.cc:72] PJRT plugin for XPU has PJRT API version 0.33. The framework PJRT API version is 0.34.
2024-09-09 14:26:29.722651: I external/intel_xla/xla/stream_executor/sycl/sycl_gpu_runtime.cc:134] Selected platform: Intel(R) Level-Zero
2024-09-09 14:26:29.722852: I external/intel_xla/xla/stream_executor/sycl/sycl_gpu_runtime.cc:159] number of sub-devices is zero, expose root device.
2024-09-09 14:26:29.723897: I external/xla/xla/service/service.cc:168] XLA service 0x625b82e02210 initialized for platform SYCL (this does not guarantee that XLA will be used). Devices:
2024-09-09 14:26:29.723932: I external/xla/xla/service/service.cc:176] StreamExecutor device (0): Intel(R) Graphics [0x7d55],
2024-09-09 14:26:29.725062: I itex/core/devices/gpu/itex_gpu_runtime.cc:130] Selected platform: Intel(R) Level-Zero
2024-09-09 14:26:29.725242: I itex/core/devices/gpu/itex_gpu_runtime.cc:155] number of sub-devices is zero, expose root device.
```

**Comment by wangkl2** [comment]
@danielgalek Since it passes the sanity check now, could you please provide a minimum example reproducer for us to reproduce your core dumped issue?

**Comment by wangkl2** [comment]
@danielgalek 
This might not be a common issue for all the workloads. I can successfully run some samples under https://github.com/intel/intel-extension-for-tensorflow/tree/main/examples on Intel XPU. But I can reproduce your `Check failed: IsAligned()` and core dumped issue with the code snippet for the LSTM training workload you provided. 

Some info from the backtrace with `gdb` on the core dumped file:
```
#7  0x00007fffb60795fc in tensorflow::Tensor::CheckTypeAndIsAligned(tensorflow::DataType) const () from /hd2/wangk2/miniforge3/envs/itex-xpu-py310/lib/python3.10/site-packages/tensorflow/python/platform/../../libtensorflow_framework.so.2                                                                                                       
#8  0x00007fffa89ddce7 in void tensorflow::CopyTensor<Eigen::ThreadPoolDevice, float>(tensorflow::OpKernelContext*, tensorflow::Tensor&, tensorflow::Tensor&) () from /hd2/wangk2/miniforge3/envs/itex-xpu-py310/lib/python3.10/site-packages/tensorflow/python/platform/../../libtensorflow_cc.so.2                                                                                                              
#9  0x00007fffa89dd658 in tensorflow::TensorListFromTensor<Eigen::ThreadPoolDevice, float>::Compute(tensorflow::OpKernelContext*) () from /hd2/wangk2/miniforge3/envs/itex-xpu-py310/lib/python3.10/site-packages/tensorflow/python/platform/../../libtensorflow_cc.so.2
#10 0x00007fffb2302987 in tensorflow::NextPluggableDevice::Compute(tensorflow::OpKernel*, tensorflow::OpKernelContext*)       () from /hd2/wangk2/miniforge3/envs/itex-xpu-py310/lib/python3.10/site-packages/tensorflow/python/platform/../../libtensorflow_cc.so.2
```

Will look into this and give feedback to you later.

**Comment by danielgalek** [comment]
Thank you very much @wangkl2, I appreciate your help and continous support! I was also able to run the examples from github successfully, which is why I was surprised that I was encountering the issue with my code. I first assumed I just had too much data but as you saw in my sample code, even with a small sample, something is not going well. My guess is it might be related to the sliding windows for LSTM but that could be a cooincidence.

**Comment by wangkl2** [comment]
@danielgalek 
As shown in the stack trace of gdb, it crashed with the `Check failed: IsAligned()` issue during calling `tensorflow::Tensor::CheckTypeAndIsAligned()` after calling `tensorflow::TensorListFromTensor` at `tensorflow/python/platform/../../libtensorflow_framework.so.2`
```
#7  0x00007fffb60795fc in tensorflow::Tensor::CheckTypeAndIsAligned(tensorflow::DataType) const () from /hd2/wangk2/miniforge3/envs/itex-xpu-py310/lib/python3.10/site-packages/tensorflow/python/platform/../../libtensorflow_framework.so.2                                                                                                       
#8  0x00007fffa89ddce7 in void tensorflow::CopyTensor<Eigen::ThreadPoolDevice, float>(tensorflow::OpKernelContext*, tensorflow::Tensor&, tensorflow::Tensor&) () from /hd2/wangk2/miniforge3/envs/itex-xpu-py310/lib/python3.10/site-packages/tensorflow/python/platform/../../libtensorflow_cc.so.2                                                                                                              
#9  0x00007fffa89dd658 in tensorflow::TensorListFromTensor<Eigen::ThreadPoolDevice, float>::Compute(tensorflow::OpKernelContext*) () from /hd2/wangk2/miniforge3/envs/itex-xpu-py310/lib/python3.10/site-packages/tensorflow/python/platform/../../libtensorflow_cc.so.2
#10 0x00007fffb2302987 in tensorflow::NextPluggableDevice::Compute(tensorflow::OpKernel*, tensorflow::OpKernelContext*)       () from /hd2/wangk2/miniforge3/envs/itex-xpu-py310/lib/python3.10/site-packages/tensorflow/python/platform/../../libtensorflow_cc.so.2
```

However, the `TensorList` related ops are not supported with the `NextPluggableDevice` by Google TF 2.15, which is listed in the `Known Issues` section in the [ITEX v2.15.0.1 release notes](https://github.com/intel/intel-extension-for-tensorflow/releases/tag/v2.15.0.1). The `NextPluggableDevice` mechanism is now the default plugin (to support XLA), see more details in [Next-Pluggable-Device](https://github.com/intel/intel-extension-for-tensorflow/blob/main/docs/guide/next_pluggable_device.md). You can switch back to the original `PluggableDevice` device plugin during the runtime before running your LSTM workload, which supports `TensorList` module.

To enable `PluggableDevice`, use the following command: 
```
export ITEX_ENABLE_NEXTPLUGGABLE_DEVICE=0
```

I've verified that your code could now smoothly execute training on Arc GPU with the `PluggableDevice` plugin, with the following output:
```
2024-09-24 15:49:00.256409: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:272] Created TensorFlow device (/job:localhost/replica:0/task:0/device:XPU:0 with 0 MB memory) -> physical PluggableDevice (device: 0, name: XPU, pci bus id: <undefined>)
2024-09-24 15:49:00.256697: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:272] Created TensorFlow device (/job:localhost/replica:0/task:0/device:XPU:1 with 0 MB memory) -> physical PluggableDevice (device: 1, name: XPU, pci bus id: <undefined>)
Epoch 1/10
2024-09-24 15:49:00.853050: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:117] Plugin optimizer for device_type XPU is enabled.
49/49 [==============================] - 1s 12ms/step - loss: 0.1606
Epoch 2/10
49/49 [==============================] - 1s 11ms/step - loss: 0.0942
Epoch 3/10
49/49 [==============================] - 1s 12ms/step - loss: 0.0917
Epoch 4/10
49/49 [==============================] - 1s 11ms/step - loss: 0.0912
Epoch 5/10
49/49 [==============================] - 1s 10ms/step - loss: 0.0938
Epoch 6/10
49/49 [==============================] - 1s 11ms/step - loss: 0.0907
Epoch 7/10
49/49 [==============================] - 1s 13ms/step - loss: 0.0901
Epoch 8/10
49/49 [==============================] - 1s 11ms/step - loss: 0.0891
Epoch 9/10
49/49 [==============================] - 1s 12ms/step - loss: 0.0887
Epoch 10/10
49/49 [==============================] - 1s 14ms/step - loss: 0.0879
Model training complete
```

Please check it out on you side.

**Comment by danielgalek** [comment]
Thank you very much for looking into this @wangkl2 ! 

I missed to make the connection between the issue and the missing support for `TensorLists`. Thank you for pointing it out and for providing a fix. I can confirm that my sample code and my original code run through successfully when switching to `PluggableDevice` and are using the Arc's GPU.

Would you be in a position to share if and approx. when support for `TensorList` with `NextPluggableDevice` could be expected?

**Comment by wangkl2** [comment]
@danielgalek Great to see it also works now on your side!
For the `TensorList` support needed by LSTM and other models, 

a) If XLA JIT compilation is not needed for your case, there is nearly no difference between choosing `PluggableDevice` and `NextPluggableDevice`. Then you should be good to use `PluggableDevice` with the solution I've provided above.

b) If you want to use XLA JIT, the development of `TensorList` related feature with the `NextPluggableDeivce` plugin depends on and is pending on the support of the PRJT API (by the Google team), such as the support of the split/slice operations on the PJRT buffer. But we're not sure when it will be supported. Please stay tuned.

**Comment by danielgalek** [comment]
You have been a great help @wangkl2  thank you very much for your time, support and great project. I will stay tuned!
